modules/homebrew.nix

```diff
@@ -1,219 +1,751 @@
+          <command>brew uninstall</command> is run for those formulae.
+
+          When set to <literal>"zap"</literal>, <command>nix-darwin</command> invokes
+          <command>brew bundle [install]</command> with the <command>--cleanup --zap</command>
+          flags. This uninstalls all formulae not listed in the generated Brewfile, and if the
+          formula is a cask, removes all files associated with that cask. In other words,
+          <command>brew uninstall --zap</command> is run for all those formulae.
+
+          If you plan on exclusively using <command>nix-darwin</command> to manage formulae
+          installed by Homebrew, you probably want to set this option to
+          <literal>"uninstall"</literal> or <literal>"zap"</literal>.
+        '';
+      };
+      update = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Whether to enable Homebrew to auto-update itself and all formulae during
+          <command>nix-darwin</command> system activation. The default is <literal>false</literal>
+          so that repeated invocations of <command>darwin-rebuild switch</command> are idempotent.
+
+          If you prefer this behavior you many also want to consider enabling
+          ${mkDocOptionLink "homebrew.global.noAutoUpdate"}.
+        '';
+      };
+      upgrade = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Whether to enable Homebrew to upgrade outdated formulae and Mac App Store apps during
+          <command>nix-darwin</command> system activation. The default is <literal>false</literal>
+          so that repeated invocations of <command>darwin-rebuild switch</command> are idempotent.
+        '';
+      };
+
+      brewBundleCmd = mkInternalOption { type = types.str; };
+    };
+
+    config = {
+      brewBundleCmd = concatStringsSep " " (
+        optional (!config.update) "HOMEBREW_NO_AUTO_UPDATE=1"
+        ++ [ "brew bundle --file='${brewfileFile}' --no-lock" ]
+        ++ optional (!config.upgrade) "--no-upgrade"
+        ++ optional (config.cleanup == "uninstall" || config.cleanup == "zap") "--cleanup"
+        ++ optional (config.cleanup == "zap") "--zap"
+      );
+    };
+  };
+
+  globalOptions = { config, ... }: {
+    options = {
+      brewfile = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Whether to enable Homebrew to automatically use the Brewfile that this module generates in
+          the Nix store, when you manually invoke <command>brew bundle</command>.
+
+          Implementation note: when enabled, this option sets the
+          <literal>HOMEBREW_BUNDLE_FILE</literal> environment variable to the path of the Brewfile
+          that this module generates in the Nix store, by adding it to
+          ${mkDocOptionLink "environment.variables"}.
+        '';
+      };
+      noAutoUpdate = mkOption {
```

Comment:
Are we sure the double negative here won't be confusing? Consider flipping the default and dropping the "no."

Reply:
I'm torn about this. Here's a brain dump of my current thinking:

* When I'm designing am module, I want it's default behavior to be what most people would expect, e.g., not cause surprising or unexpected changes to how their system works.
* I like to add boolean options as toggles to enable some additional/different behavior/functionality/setting that folks would want to explicitly opt-in to, where the default is that it's off (false), and so the user setting the option to true is them saying "I want this additional behavior/functionality/setting enabled".
* That said, I'm definitely not dogmatic about this. I have no love for double negatives, and so I sometimes make an exception from the above when it's needlessly confusing.
* In this particular case I'm currently leaning against removing the double negative because:
  * I like that, for folks that are familiar with Homebrew, the option names line up with the with names of the environment variables they'd be familiar with and be looking to set.
  * The `homebrew.global.noLock` option has existed since the module was created. I'm  aware that I could use `mkChangedOptionModule` to rename the option while preserving backwards compatibility with the current option, but I don't love that folks would have to notice that they should swap the value when renaming the option to address the warning they'll see when trying to build their config after this update. I guess I could use `doRename` directly, and provide a custom warning message, but I have a reflex against doing that since it will clutter up the module a little more. (Maybe I should just get over that thought ¯\_(ツ)_/¯.)

Reply:
That said, I don't feel very strongly about this, and so if most folks think this is an obvious improvement, I'd be happy to defer to the wisdom of the crowd and make the changes.

Reply:
I'm leaving this up to your discretion. The default doesn't have to change just because the name changes though.

**My thoughts on the topic**:

I don't feel like modules should follow UI choices of CLIs, they're fundamentally different interfaces. Adding `--no-lock` unambiguously means "don't lock," setting `noLock = true` does not share this clear-cut interpretation.

Another way to accomodate experienced users and new users alike is to offer both options and asserting they are each others' opposite if both are set. Not elegant but it covers both concerns, inexperienced people  aren't forced into boolean contortions and experienced people don't have to unlearn their domain knowledge.

Reply:
I'm feeling more convinced that this change should be made. I'll play around with this today.